File: apps/web/src/lib/conversion-api.ts
import {
  getCanonicalFileName,
  getConversionPair,
  isConversionPairEnabled,
  type FormatId,
} from "@/lib/formats";

export type ConversionStatus =
  | "idle"
  | "ready"
  | "uploading"
  | "queued"
  | "starting"
  | "converting"
  | "completed"
  | "failed";

export interface CreateConversionRequest {
  source_format: FormatId;
  target_format: FormatId;
}

export interface CreateConversionResponse {
  conversion_id: string;
  status: "completed";
  download_url: string;
  output_key?: string;
  content_type?: string;
  size?: number;
}

interface CreateUploadResponse {
  upload_id: string;
  object_key: string;
  upload_url: string;
  content_type: string;
  expires_in: number;
}

interface QueueConversionResponse {
  conversion_id: string;
  source_format: FormatId;
  target_format: FormatId;
  input_key: string;
  status: "queued";
}

interface ConversionStatusResponse {
  conversion_id: string;
  status: "processing" | "completed" | "failed";
  message?: string;
  output_key?: string;
  content_type?: string;
  size?: number;
  download_url?: string;
  download_expires_in?: number;
}

export type SubmissionReadiness =
  | { ready: true; mode: "live" }
  | {
      ready: false;
      reason: "api-missing";
      message: string;
    };

export class ConversionApiError extends Error {
  readonly retryable: boolean;

  constructor(message: string, retryable = false) {
    super(message);
    this.name = "ConversionApiError";
    this.retryable = retryable;
  }
}

const apiBaseUrl = process.env.NEXT_PUBLIC_CONVERTIX_API_URL?.trim().replace(
  /\/$/,
  "",
);

export function getSubmissionReadiness(): SubmissionReadiness {
  if (!apiBaseUrl) {
    return {
      ready: false,
      reason: "api-missing",
      message: "The conversion service is unavailable right now.",
    };
  }

  return {
    ready: true,
    mode: "live",
  };
}

function parseApiError(
  response: Response,
  fallback: string,
): ConversionApiError {
  const retryable = response.status >= 500 || response.status === 429;
  return new ConversionApiError(fallback, retryable);
}

function wait(ms: number, signal?: AbortSignal): Promise<void> {
  return new Promise((resolve, reject) => {
    if (signal?.aborted) {
      reject(new DOMException("Aborted", "AbortError"));
      return;
    }

    const timeout = window.setTimeout(() => {
      signal?.removeEventListener("abort", abort);
      resolve();
    }, ms);

    function abort() {
      window.clearTimeout(timeout);
      reject(new DOMException("Aborted", "AbortError"));
    }

    signal?.addEventListener("abort", abort, { once: true });
  });
}

export type PdfCompressionLevel = "light" | "balanced" | "maximum";

export interface CreateConversionRequest {
  source_format: FormatId;
  target_format: FormatId;
  compression_level?: PdfCompressionLevel;
}

export async function createConversion(
  file: File,
  request: CreateConversionRequest,
  signal?: AbortSignal,
  onStatus?: (status: ConversionStatus) => void,
): Promise<CreateConversionResponse> {
  const readiness = getSubmissionReadiness();

  if (!readiness.ready) {
    throw new ConversionApiError(readiness.message);
  }

  const isPdfCompression =
    request.source_format === "pdf" &&
    request.target_format === "pdf" &&
    request.compression_level !== undefined;

  if (!isPdfCompression) {
    const pair = getConversionPair(
      request.source_format,
      request.target_format,
    );

    if (!pair || !isConversionPairEnabled(pair)) {
      throw new ConversionApiError(
        "Convertix can’t run that conversion yet. Choose another output format.",
      );
    }
  }

  const browserContentType = file.type || "application/octet-stream";

  onStatus?.("uploading");

  let uploadResponse: Response;

  try {
    uploadResponse = await fetch(`${apiBaseUrl}/uploads`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        filename: getCanonicalFileName(file.name, request.source_format),
        content_type: browserContentType,
      }),
      signal,
    });
  } catch (error) {
    if (error instanceof DOMException && error.name === "AbortError") {
      throw error;
    }

    throw new ConversionApiError(
      "We couldn’t start the upload. Check your connection and try again.",
      true,
    );
  }

  if (!uploadResponse.ok) {
    throw await parseApiError(
      uploadResponse,
      "Convertix couldn’t prepare the upload. Try again.",
    );
  }

  const upload = (await uploadResponse.json()) as Partial<CreateUploadResponse>;

  if (!upload.upload_url || !upload.object_key) {
    throw new ConversionApiError(
      "Convertix couldn’t prepare the upload. Try again.",
      true,
    );
  }

  const uploadContentType = upload.content_type || browserContentType;

  let s3Response: Response;

  try {
    s3Response = await fetch(upload.upload_url, {
      method: "PUT",
      headers: {
        "Content-Type": uploadContentType,
      },
      body: file,
      signal,
    });
  } catch (error) {
    if (error instanceof DOMException && error.name === "AbortError") {
      throw error;
    }

    throw new ConversionApiError(
      "We couldn’t upload your file. The browser may have blocked the upload or the connection was interrupted. Try again.",
      true,
    );
  }

  if (!s3Response.ok) {
    throw new ConversionApiError(
      "The upload reached storage but didn’t complete successfully. Try again.",
      true,
    );
  }

  let queueResponse: Response;

  try {
    queueResponse = await fetch(`${apiBaseUrl}/conversions`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        source_format: request.source_format,
        target_format: request.target_format,
        input_key: upload.object_key,
        ...(request.compression_level
          ? { compression_level: request.compression_level }
          : {}),
      }),
      signal,
    });
  } catch (error) {
    if (error instanceof DOMException && error.name === "AbortError") {
      throw error;
    }

    throw new ConversionApiError(
      "We couldn’t start the conversion. Try again.",
      true,
    );
  }

  if (!queueResponse.ok) {
    throw await parseApiError(
      queueResponse,
      "Convertix couldn’t start the conversion. Try again.",
    );
  }

  const queued =
    (await queueResponse.json()) as Partial<QueueConversionResponse>;

  if (!queued.conversion_id) {
    throw new ConversionApiError(
      "Convertix couldn’t start the conversion. Try again.",
      true,
    );
  }

  onStatus?.("queued");

  const startedAt = Date.now();
  const timeoutMs = 15 * 60 * 1000;

  while (Date.now() - startedAt < timeoutMs) {
    await wait(3000, signal);

    let statusResponse: Response;

    try {
      statusResponse = await fetch(
        `${apiBaseUrl}/conversions/${queued.conversion_id}`,
        {
          method: "GET",
          signal,
          cache: "no-store",
        },
      );
    } catch (error) {
      if (error instanceof DOMException && error.name === "AbortError") {
        throw error;
      }

      // A temporary polling failure should not kill a conversion that
      // may still be running successfully in AWS.
      continue;
    }

    if (!statusResponse.ok) {
      if (statusResponse.status >= 500 || statusResponse.status === 429) {
        continue;
      }

      throw await parseApiError(
        statusResponse,
        "Convertix lost track of the conversion. Try again.",
      );
    }

    const status =
      (await statusResponse.json()) as Partial<ConversionStatusResponse>;

    if (status.status === "processing") {
      const elapsed = Date.now() - startedAt;

      if (elapsed < 60_000) {
        onStatus?.("starting");
      } else {
        onStatus?.("converting");
      }

      continue;
    }

    if (status.status === "failed") {
      onStatus?.("failed");
      throw new ConversionApiError(
        status.message ??
          "Convertix couldn’t convert this file. Check that the file is valid and try again.",
        false,
      );
    }

    if (status.status === "completed") {
      if (!status.download_url) {
        throw new ConversionApiError(
          "Your file was converted, but the download link is missing. Try the conversion again.",
          true,
        );
      }

      onStatus?.("completed");

      return {
        conversion_id: queued.conversion_id,
        status: "completed",
        download_url: status.download_url,
        output_key: status.output_key,
        content_type: status.content_type,
        size: status.size,
      };
    }
  }

  throw new ConversionApiError(
    "This conversion has taken longer than 15 minutes. Try again.",
    true,
  );
}

File: apps/apple/Convertix/Services/ConversionAPI.swift
import Foundation
import UniformTypeIdentifiers

struct ConversionAPI: Sendable {
    private struct UploadRequest: Encodable {
        let filename: String
        let contentType: String

        enum CodingKeys: String, CodingKey {
            case filename
            case contentType = "content_type"
        }
    }

    private struct UploadResponse: Decodable {
        let uploadID: UUID
        let objectKey: String
        let uploadURL: URL
        let contentType: String
        let expiresIn: Int

        enum CodingKeys: String, CodingKey {
            case uploadID = "upload_id"
            case objectKey = "object_key"
            case uploadURL = "upload_url"
            case contentType = "content_type"
            case expiresIn = "expires_in"
        }
    }

    private struct QueueRequest: Encodable {
        let sourceFormat: String
        let targetFormat: String
        let inputKey: String

        enum CodingKeys: String, CodingKey {
            case sourceFormat = "source_format"
            case targetFormat = "target_format"
            case inputKey = "input_key"
        }
    }

    private struct QueueResponse: Decodable {
        let conversionID: UUID
        let status: String

        enum CodingKeys: String, CodingKey {
            case conversionID = "conversion_id"
            case status
        }
    }

    private struct StatusResponse: Decodable {
        let conversionID: UUID
        let status: String
        let message: String?
        let outputKey: String?
        let contentType: String?
        let size: Int64?
        let downloadURL: URL?
        let downloadExpiresIn: Int?

        enum CodingKeys: String, CodingKey {
            case conversionID = "conversion_id"
            case status
            case message
            case outputKey = "output_key"
            case contentType = "content_type"
            case size
            case downloadURL = "download_url"
            case downloadExpiresIn = "download_expires_in"
        }
    }

    private struct ErrorResponse: Decodable {
        let error: String?
    }

    private let session: URLSession
    private let baseURL: URL
    private let pollInterval: Duration
    private let timeout: TimeInterval
    private let maximumConsecutivePollingFailures: Int

    init(
        session: URLSession = .shared,
        configuration: AppConfiguration,
        pollInterval: Duration = .seconds(3),
        timeout: TimeInterval = 15 * 60,
        maximumConsecutivePollingFailures: Int = 5
    ) {
        self.session = session
        baseURL = configuration.apiBaseURL
        self.pollInterval = pollInterval
        self.timeout = timeout
        self.maximumConsecutivePollingFailures = maximumConsecutivePollingFailures
    }

    init(session: URLSession = .shared, bundle: Bundle = .main) throws {
        self.init(session: session, configuration: try AppConfiguration(bundle: bundle))
    }

    func convert(
        fileURL: URL,
        route: ConversionRoute,
        onStatus: @escaping @MainActor (ConversionStatus) -> Void
    ) async throws -> ConversionResult {
        let sourceFormat = route.source.lowercased()
        let targetFormat = route.target.lowercased()
        try validate(fileURL: fileURL, sourceFormat: sourceFormat)

        await onStatus(.uploading)
        let contentType = UTType(filenameExtension: fileURL.pathExtension)?.preferredMIMEType
            ?? "application/octet-stream"
        let canonicalFilename = canonicalFilename(
            originalName: fileURL.lastPathComponent,
            sourceFormat: sourceFormat
        )
        let upload: UploadResponse = try await sendJSON(
            UploadRequest(filename: canonicalFilename, contentType: contentType),
            to: endpoint("uploads")
        )

        var uploadRequest = URLRequest(url: upload.uploadURL)
        uploadRequest.httpMethod = "PUT"
        uploadRequest.setValue(upload.contentType, forHTTPHeaderField: "Content-Type")

        do {
            let (_, response) = try await session.upload(for: uploadRequest, fromFile: fileURL)
            try validate(response)
        } catch is CancellationError {
            throw CancellationError()
        } catch let error as ConversionAPIError {
            throw error
        } catch {
            throw ConversionAPIError.network(
                "We couldn’t upload your file. Check your connection and try again."
            )
        }

        await onStatus(.queued)
        let queued: QueueResponse = try await sendJSON(
            QueueRequest(
                sourceFormat: sourceFormat,
                targetFormat: targetFormat,
                inputKey: upload.objectKey
            ),
            to: endpoint("conversions")
        )

        guard queued.status == "queued" else {
            throw ConversionAPIError.invalidResponse
        }

        let deadline = Date().addingTimeInterval(timeout)
        var consecutivePollingFailures = 0
        while Date() < deadline {
            try await Task.sleep(for: pollInterval)
            let status: StatusResponse

            do {
                status = try await get(endpoint("conversions/\(queued.conversionID.uuidString)"))
            } catch let error as ConversionAPIError where error.isRetryable {
                consecutivePollingFailures += 1
                guard consecutivePollingFailures < maximumConsecutivePollingFailures else {
                    throw error
                }
                continue
            }

            consecutivePollingFailures = 0

            switch status.status.lowercased() {
            case "processing":
                await onStatus(.processing)
            case "completed":
                guard let downloadURL = status.downloadURL else {
                    throw ConversionAPIError.missingDownloadURL
                }
                let result = ConversionResult(
                    conversionID: status.conversionID,
                    downloadURL: downloadURL,
                    outputKey: status.outputKey,
                    contentType: status.contentType,
                    outputSize: status.size,
                    suggestedFilename: outputFilename(
                        originalName: fileURL.lastPathComponent,
                        targetFormat: targetFormat
                    )
                )
                await onStatus(.completed(downloadURL))
                return result
            case "failed":
                throw ConversionAPIError.conversionFailed(
                    status.message
                        ?? "Convertix couldn’t convert this file. Check that it is valid and try again."
                )
            default:
                throw ConversionAPIError.invalidResponse
            }
        }

        throw ConversionAPIError.timedOut
    }

    func download(_ result: ConversionResult) async throws -> URL {
        let (temporaryURL, response) = try await session.download(from: result.downloadURL)
        try validate(response)

        let directory = FileManager.default.temporaryDirectory
            .appending(path: "Convertix", directoryHint: .isDirectory)
        try FileManager.default.createDirectory(
            at: directory,
            withIntermediateDirectories: true
        )
        let destination = directory.appending(path: result.suggestedFilename)
        if FileManager.default.fileExists(atPath: destination.path) {
            try FileManager.default.removeItem(at: destination)
        }
        try FileManager.default.moveItem(at: temporaryURL, to: destination)
        return destination
    }

    private func canonicalFilename(originalName: String, sourceFormat: String) -> String {
        let stem = URL(fileURLWithPath: originalName)
            .deletingPathExtension()
            .lastPathComponent
        let canonicalExtension = sourceFormat == "jpg" ? "jpg" : sourceFormat
        return "\(stem.isEmpty ? "upload" : stem).\(canonicalExtension)"
    }

    private func outputFilename(originalName: String, targetFormat: String) -> String {
        let stem = URL(fileURLWithPath: originalName)
            .deletingPathExtension()
            .lastPathComponent
        return "\(stem.isEmpty ? "convertix" : stem)-converted.\(targetFormat)"
    }

    private func validate(fileURL: URL, sourceFormat: String) throws {
        let extensionFormat = fileURL.pathExtension.lowercased()
        let acceptedExtensions: Set<String>
        switch sourceFormat {
        case "jpg":
            acceptedExtensions = ["jpg", "jpeg"]
        case "heic":
            acceptedExtensions = ["heic"]
        case "heif":
            acceptedExtensions = ["heif"]
        default:
            acceptedExtensions = [sourceFormat]
        }

        guard acceptedExtensions.contains(extensionFormat) else {
            throw ConversionAPIError.formatMismatch(expected: sourceFormat.uppercased())
        }

        if let size = try fileURL.resourceValues(forKeys: [.fileSizeKey]).fileSize {
            guard size > 0 else {
                throw ConversionAPIError.emptyFile
            }
            guard size <= 100 * 1024 * 1024 else {
                throw ConversionAPIError.fileTooLarge
            }
        }
    }

    private func endpoint(_ path: String) -> URL {
        baseURL.appending(path: path)
    }

    private func sendJSON<Request: Encodable, Response: Decodable>(
        _ body: Request,
        to url: URL
    ) async throws -> Response {
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.timeoutInterval = 30
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONEncoder.convertix.encode(body)
        return try await perform(request)
    }

    private func get<Response: Decodable>(_ url: URL) async throws -> Response {
        var request = URLRequest(url: url)
        request.cachePolicy = .reloadIgnoringLocalCacheData
        request.timeoutInterval = 30
        return try await perform(request)
    }

    private func perform<Response: Decodable>(_ request: URLRequest) async throws -> Response {
        let data: Data
        let response: URLResponse

        do {
            (data, response) = try await session.data(for: request)
        } catch is CancellationError {
            throw CancellationError()
        } catch {
            throw ConversionAPIError.network(
                "Convertix couldn’t reach the conversion service. Check your connection and try again."
            )
        }

        try validate(response, body: data)

        do {
            return try JSONDecoder.convertix.decode(Response.self, from: data)
        } catch {
            throw ConversionAPIError.invalidResponse
        }
    }

    private func validate(_ response: URLResponse, body: Data = Data()) throws {
        guard let httpResponse = response as? HTTPURLResponse else {
            throw ConversionAPIError.invalidResponse
        }
        guard (200..<300).contains(httpResponse.statusCode) else {
            let code = try? JSONDecoder.convertix.decode(ErrorResponse.self, from: body).error
            throw ConversionAPIError.server(status: httpResponse.statusCode, code: code)
        }
    }
}

extension JSONDecoder {
    static var convertix: JSONDecoder {
        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        return decoder
    }
}

extension JSONEncoder {
    static var convertix: JSONEncoder {
        let encoder = JSONEncoder()
        encoder.dateEncodingStrategy = .iso8601
        return encoder
    }
}

enum ConversionAPIError: LocalizedError, Equatable {
    case conversionFailed(String)
    case emptyFile
    case fileTooLarge
    case formatMismatch(expected: String)
    case invalidResponse
    case missingDownloadURL
    case network(String)
    case server(status: Int, code: String?)
    case timedOut

    var isRetryable: Bool {
        switch self {
        case .network, .missingDownloadURL:
            true
        case let .server(status, _):
            status == 429 || status >= 500
        default:
            false
        }
    }

    var errorDescription: String? {
        switch self {
        case let .conversionFailed(message), let .network(message):
            message
        case .emptyFile:
            "That file is empty. Choose a file that contains something to convert."
        case .fileTooLarge:
            "Choose a file smaller than 100 MB."
        case let .formatMismatch(expected):
            "Choose a \(expected) file or select the matching source format."
        case .invalidResponse:
            "The conversion service returned an unexpected response. Try again."
        case .missingDownloadURL:
            "Your file was converted, but the download link is missing. Try again."
        case let .server(status, code):
            Self.message(for: code) ?? "The conversion service returned HTTP \(status)."
        case .timedOut:
            "This conversion has taken longer than 15 minutes. Try again."
        }
    }

    private static func message(for code: String?) -> String? {
        switch code {
        case "unsupported_upload_format":
            "Convertix doesn’t support uploading this file format."
        case "unsupported_conversion":
            "Convertix can’t run that conversion yet. Choose another output format."
        case "input_format_mismatch":
            "The selected file doesn’t match the chosen source format."
        case "storage_not_configured", "queue_not_configured":
            "The conversion service is temporarily unavailable."
        case "failed_to_create_upload_url", "failed_to_queue_conversion",
             "failed_to_check_conversion", "failed_to_create_download_url":
            "Convertix couldn’t complete that request. Try again."
        default:
            nil
        }
    }
}
